File: Server/src/main/java/com/github/dr/rwserver/net/udp/impl/Timer.java
package com.github.dr.rwserver.net.udp.impl;

public class Timer extends Thread {
    public Timer(String name, Runnable task) {
        super(name);
        setDaemon(true);

        _task = task;
        _delay = 0;
        _period = 0;
        start();
    }

    @Override
    public void run() {
        while (!_stopped) {

            synchronized (this) {

                while (!_scheduled && !_stopped) {
                    try {
                        wait();
                    }
                    catch (InterruptedException xcp) {
                        xcp.printStackTrace();
                    }
                }

                if (_stopped) {
                    break;
                }
            }

            synchronized (_lock) {

                _reset = false;
                _canceled = false;

                if (_delay > 0) {
                    try {
                        _lock.wait(_delay);
                    }
                    catch (InterruptedException xcp) {
                        xcp.printStackTrace();
                    }
                }

                if (_canceled) {
                    continue;
                }
            }

            if (!_reset) {
                _task.run();
            }

            if (_period > 0) {

                while (true) {

                    synchronized (_lock) {

                        _reset = false;

                        try {
                            _lock.wait(_period);
                        }
                        catch (InterruptedException xcp) {
                            xcp.printStackTrace();
                        }

                        if (_canceled) {
                            break;
                        }

                        if (_reset) {
                            continue;
                        }
                    }

                    _task.run();

                }
            }
        }
    }

    public synchronized void schedule(long delay) {
        schedule(delay, 0);
    }

    public synchronized void schedule(long delay, long period) {
        _delay = delay;
        _period = period;


        if (_scheduled) {
            throw new IllegalStateException("already scheduled");
        }

        _scheduled = true;
        notify();

        synchronized (_lock) {
            _lock.notify();
        }
    }

    public synchronized boolean isScheduled() {
        return _scheduled;
    }

    public synchronized boolean isIdle() {
        return !isScheduled();
    }

    public synchronized void reset() {
        synchronized (_lock) {
            _reset = true;
            _lock.notify();
        }
    }

    public synchronized void cancel() {
        _scheduled = false;
        synchronized (_lock) {
            _canceled = true;
            _lock.notify();
        }
    }

    public synchronized void destroy() {
        cancel();
        _stopped = true;
        notify();
    }

    private final Runnable _task;
    private long     _delay;
    private long     _period;
    private boolean  _canceled;
    private boolean  _scheduled;
    private boolean  _reset;
    private boolean  _stopped;
    private final Object   _lock = new Object();
}

File: Server-Core/src/main/java/net/rwhps/server/util/internal/net/rudp/impl/Timer.kt
/*
 * Copyright 2020-2024 RW-HPS Team and contributors.
 *
 * 此源代码的使用受 GNU AFFERO GENERAL PUBLIC LICENSE version 3 许可证的约束, 可以在以下链接找到该许可证.
 * Use of this source code is governed by the GNU AGPLv3 license that can be found through the following link.
 *
 * https://github.com/RW-HPS/RW-HPS/blob/master/LICENSE
 */

package net.rwhps.server.util.internal.net.rudp.impl

import java.util.concurrent.TimeUnit
import java.util.concurrent.locks.Condition
import java.util.concurrent.locks.ReentrantLock

/**
 * Timer
 *
 * @author Dr (dr@der.kim)
 */
class Timer(private val name: String, private val task: Runnable) {
    private val lock = ReentrantLock()
    private val condition: Condition = lock.newCondition()
    private var delay: Long = 0
    private var period: Long = 0
    private var canceled: Boolean = false
    private var scheduled: Boolean = false
    private var reset: Boolean = false
    private var stopped: Boolean = false

    fun start() {
        Thread {
            while (!Thread.currentThread().isInterrupted) {
                lock.lock()
                try {
                    while (!scheduled) {
                        condition.await()
                    }
                    if (stopped) {
                        break
                    }
                    if (delay > 0) {
                        if (condition.await(delay, TimeUnit.MILLISECONDS)) {
                            if (canceled) {
                                continue
                            }
                        } else {
                            if (canceled) {
                                continue
                            }
                            task.run()
                        }
                    } else {
                        if (canceled) {
                            continue
                        }
                        task.run()
                    }
                    if (period > 0) {
                        while (true) {
                            if (reset) {
                                break
                            }
                            if (condition.await(period, TimeUnit.MILLISECONDS)) {
                                if (canceled) {
                                    break
                                }
                                if (!reset) {
                                    task.run()
                                }
                            } else {
                                if (canceled) {
                                    break
                                }
                                if (!reset) {
                                    task.run()
                                }
                            }
                        }
                    }
                } finally {
                    lock.unlock()
                }
            }
        }.start()
    }

    @Synchronized
    fun schedule(delay: Long) {
        schedule(delay, 0)
    }

    @Synchronized
    fun schedule(delay: Long, period: Long) {
        this.delay = delay
        this.period = period
        if (scheduled) {
            throw IllegalStateException("already scheduled")
        }
        scheduled = true
        lock.lock()
        try {
            condition.signalAll()
        } finally {
            lock.unlock()
        }
    }

    @Synchronized
    fun isScheduled(): Boolean {
        return scheduled
    }

    @Synchronized
    fun isIdle(): Boolean {
        return !isScheduled()
    }

    @Synchronized
    fun reset() {
        lock.lock()
        try {
            reset = true
        } finally {
            lock.unlock()
        }
    }

    @Synchronized
    fun cancel() {
        lock.lock()
        try {
            canceled = true
        } finally {
            lock.unlock()
        }
    }

    @Synchronized
    fun destroy() {
        cancel()
        stopped = true
    }
}